app/src/main/java/com/amaze/filemanager/asynchronous/services/AbstractProgressiveService.java

```diff
@@ -240,14 +241,16 @@ private void publishCompletedResult(int id1) {
   }
 
   protected void addFirstDatapoint(String name, int amountOfFiles, long totalBytes, boolean move) {
-    if (!getDataPackages().isEmpty()) {
-      throw new IllegalStateException("This is not the first datapoint!");
+    synchronized (getDataPackages()) {
```

Comment:
Why the synchronized? Why isn't this using an actual lock if its needed? Or a concurrent list?

Reply:
I doubt even concurrent list can't address the issue here. If line 247 is executed and before putting first data package in 252 it's possible someone else can add another data package again. Hence the synchronized block. I'm not sure what do you mean by the 'actual lock'

Reply:
Every usage of the list should be syncronized for this to work.

Reply:
Syncronization on objects is currently not enforced either statically or in runtime, that's why I proposed using an actual lock or the concurrent list.

Reply:
Concurrent list will not help here because after clearing the list we want to add new element to it, and before reaching the line of code we want to avoid someone else to add to the list first (in `addDatapoint`). 
I do agree that we need synchronization on every usage, so i'll be putting synchronized on both `addFirstDatapoint` and `addDatapoint`. 

All this said, the real culprit here was `clearDatapackages` which wasn't called after the extract service's onDestroy, so the list already had elements on subsequent calls to extract service. I've added clear datapackages in abstract service's onDestroy. This race condition should almost never happen now, but regardless, I'll put synchronized on both  the methods.

Reply:
1.`onDestroy` will not always be called, regardless, why does the list get presisted after the `onDestroy` is called?
2. If adding `clearDatapackages` to `onDestroy` fixes the issue and then syncronization is not needed, syncronization should be removed.
3. Otherwise you have to make sure the subclass doesn't do things with the list while it is synchronized in `AbstractProgressiveService` which could mean moving the list up to `AbstractProgressiveService` and synchronizing *on the list* not on the class object.

Reply:
1. I'm not sure, but this is the only logical explanation behind this crash. Also,  clearing it in onDestroy doesn't harm either.
2. Point 1 as well as there *might be* a small chance a race condition can occur, again, I don't see any harm in adding synchronized on these methods, because clearly the methods are adding the first data point, we wouldn't want a race condition that during it cleared the existing list and when it's again adding the first one, someone else doesn't modify the list.
3. Making the list synchronized isn't really necessary. We're simply putting new datapoints in the list, it's just when the list is being created for the first time that requires these changes. 
Again, we can get away without putting synchronized on these blocks but I don't see any harm in doing so.

Reply:
It doesn't seem a good idea to add things that are redundant, specially when related to threading.